jerry-main/main-unix-snapshot.c

```diff
@@ -0,0 +1,308 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string.h>
+
+#include "jerryscript.h"
+#include "jerryscript-port.h"
+
+#include "cli.h"
+
+/**
+ * Maximum size for loaded snapshots
+ */
+#define JERRY_BUFFER_SIZE (1048576)
+
+/**
+ * Standalone Jerry exit codes
+ */
+#define JERRY_STANDALONE_EXIT_CODE_OK   (0)
+#define JERRY_STANDALONE_EXIT_CODE_FAIL (1)
+
+static uint8_t input_buffer[ JERRY_BUFFER_SIZE ];
+static uint32_t output_buffer[ JERRY_BUFFER_SIZE / 4 ];
+
+/**
+ * Loading a single file into the memory.
+ *
+ * @return true - if loading is successful
+ *         false - otherwise
+ */
+static size_t
+read_file (uint8_t *input_pos_p, /**< next position in the input buffer */
+           const char *file_name) /**< file name */
+{
+  FILE *file = fopen (file_name, "r");
+
+  if (file == NULL)
+  {
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: failed to open file: %s\n", file_name);
+    return 0;
+  }
+
+  size_t max_size = (size_t) (input_buffer + JERRY_BUFFER_SIZE - input_pos_p);
+
+  size_t bytes_read = fread (input_pos_p, 1u, max_size, file);
+  fclose (file);
+
+  if (bytes_read == 0)
+  {
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: failed to read file: %s\n", file_name);
+    return 0;
+  }
+
+  if (bytes_read >= max_size)
```

Comment:
Is this correct? `fread` was called with `max_size`, so `bytes_read` cannot be larger than `max_size`. And if the file was exactly `max_size` large, why would that count as an error? IMO this does not catch the situation when the whole contents of the file did not fit in the buffer. (`feof` could be used to decide whether we were able to read the whole file, but we don't have that in jerry-libc, unfortunately.)

Reply:
This is what we can do. I don't think we loose too much with not supporting files which size is exactly max_size.

Reply:
But, as I tried to point out above, the "this is what we can do" is 1) a no-op and 2) incorrect. In the above condition, `bytes_read > max_size` can never happen, while `bytes_read == max_size` is simply unnecessary and incorrectly restrictive. Why have this check then if all it does is wrong?

Reply:
I cannot follow. If your file size is 12 bytes long, and you read 10 bytes, you get 10 bytes, but your file is partly loaded. Probably you get a syntax error, but it is not 100%. So there is a chance you compile something wrong, and get a success. Ok, we also loose 10 byte long files, but I can live with that. At least we don't need to process truncated files.

Reply:
`byte_read` comes from the statement `size_t bytes_read = fread (input_pos_p, 1u, max_size, file);`, see above. So, it cannot ever be bigger than `max_size`. See https://helpmanual.io/man3/fread/ . So, if you have 10 bytes left in the buffer, you'll call `fread (input_pos_p, 1u, 10, file)`. If your file was 12 bytes long, you'll get 10 into `bytes_read`. If your file was 10 bytes long, you'll also get 10 into `bytes_read`. You cannot tell.

At the very least, the condition should be changed to `bytes_read == max_size`.

Reply:
(For the records: neither main-unix.c nor main-unix-minimal.c try to do anything - e.g., signalling errors - with truncated input files.)

Reply:
(My note: I think truncating source files is less harmful than truncating snapshots)